src/bootstrap/config.rs

```diff
@@ -1452,8 +1452,15 @@ impl Config {
                 .map(|v| v.parse().expect("failed to parse rust.llvm-libunwind"));
 
             if let Some(ref backends) = rust.codegen_backends {
-                config.rust_codegen_backends =
-                    backends.iter().map(|s| INTERNER.intern_str(s)).collect();
+                let available_backends = vec!["llvm", "cranelift", "gcc"];
+
+                config.rust_codegen_backends = backends.iter().map(|s| {
+                    if !available_backends.contains(&s.as_str()) {
```

Comment:
Alternatively, we can handle these names without getting too worried. An example: when we combine `rustc_codegen_` with the backend name, it fails if we use `rustc_codegen_gcc`, resulting in `rustc_codegen_rustc_codegen_gcc`. However, we can fix this problem by not adding the `rustc_codegen_` prefix if it's already included in the name. So, out-of-tree backends won't be a problem in this situation. What do you think on this approach?

Reply:
I'm always a little wary of canonicalizing stuff instead of either simply accepting or rejecting, but it seems plausibly fine in this case.

Reply:
So with the latest change, now we panic if in-tree backends are used with `rustc_codegen_` prefix, and print a help message if out-of-tree backend is used with `rustc_codegen_` prefix. Each message includes the suggestion for fixing the problem.

Reply:
> Alternatively, we can handle these names without getting too worried. An example: when we combine rustc_codegen_ with the backend name, it fails if we use rustc_codegen_gcc, resulting in rustc_codegen_rustc_codegen_gcc. However, we can fix this problem by not adding the rustc_codegen_ prefix if it's already included in the name. So, out-of-tree backends won't be a problem in this situation. What do you think on this approach?

The reason I didn't follow this solution was because it was going to be too hacky(it didn't work out of the box, codegen backends were failing on the compilation)

Reply:
Thanks!